File: OsmAnd/src/net/osmand/plus/card/color/palette/ColorsPaletteAdapter.java
package net.osmand.plus.card.color.palette;

import android.annotation.SuppressLint;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

class ColorsPaletteAdapter extends RecyclerView.Adapter<ColorViewHolder> {

	private final FragmentActivity activity;
	private final IColorsPaletteUIController controller;
	private final ColorsPaletteElements paletteElements;
	private List<Integer> colors;
	private boolean nightMode;

	public ColorsPaletteAdapter(@NonNull FragmentActivity activity,
	                            @NonNull IColorsPaletteUIController controller,
	                            boolean nightMode) {
		this.activity = activity;
		this.controller = controller;
		this.colors = controller.getAllColors();
		this.nightMode = nightMode;
		paletteElements = new ColorsPaletteElements(activity, nightMode);
		setHasStableIds(true);
	}

	@SuppressLint("NotifyDataSetChanged")
	public void updateColorsList() {
		this.colors = controller.getAllColors();
		notifyDataSetChanged();
	}

	@NonNull
	@Override
	public ColorViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		View view = paletteElements.createCircleView(parent);
		return new ColorViewHolder(view);
	}

	@Override
	public void onBindViewHolder(@NonNull ColorViewHolder holder, int position) {
		int color = colors.get(position);
		boolean isSelected = controller.getSelectedColor() == color;
		paletteElements.updateColorItemView(holder.itemView, color, isSelected);
		holder.background.setOnClickListener(v -> {
			controller.onSelectColorFromPalette(color);
		});
		holder.background.setOnLongClickListener(v -> {
			controller.onColorItemLongClicked(activity, holder.background, color, nightMode);
			return false;
		});
	}

	@Override
	public int getItemCount() {
		return colors.size();
	}

	@Override
	public long getItemId(int position) {
		return colors.get(position);
	}

}

File: OsmAnd/src/net/osmand/plus/card/color/palette/main/v2/ColorsPaletteAdapter.kt
package net.osmand.plus.card.color.palette.main.v2

import android.annotation.SuppressLint
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import net.osmand.plus.R
import net.osmand.plus.card.color.palette.main.ColorsPaletteElements
import net.osmand.shared.palette.data.PaletteSortMode
import net.osmand.shared.palette.domain.PaletteItem

// TODO: rename to PaletteItemsCardAdapter
class ColorsPaletteAdapter(
	private val activity: FragmentActivity,
	private val controller: IColorsPaletteController,
	private val nightMode: Boolean
) : RecyclerView.Adapter<ColorsPaletteAdapter.ColorViewHolder>() {

	private val paletteElements = ColorsPaletteElements(activity, nightMode)
	private var items: List<PaletteItem>

	init {
		items = controller.getPaletteItems(PaletteSortMode.LAST_USED_TIME)
		// TODO: we could use DiffUtils instead of setHasStableIds
		setHasStableIds(true)
	}

	@SuppressLint("NotifyDataSetChanged")
	fun updateItemsList() {
		items = controller.getPaletteItems(PaletteSortMode.LAST_USED_TIME)
		notifyDataSetChanged()
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ColorViewHolder {
		val view = paletteElements.createCircleView(parent)
		return ColorViewHolder(view)
	}

	override fun onBindViewHolder(holder: ColorViewHolder, position: Int) {
		val item = items[position]
		val isSelected = controller.isPaletteItemSelected(item)

		// TODO: don't check here
		if (item is PaletteItem.Solid) {
			paletteElements.updateColorItemView(holder.itemView, item.color, isSelected)
		}

		holder.itemView.setOnClickListener {
			controller.onSelectItemFromPalette(item, false)
		}
		holder.itemView.setOnLongClickListener {
			controller.onPaletteItemLongClick(activity, holder.background, item, nightMode)
			false
		}
	}

	fun askNotifyItemChanged(item: PaletteItem?) {
		if (item != null) {
			val index = indexOf(item)
			if (index >= 0) {
				notifyItemChanged(index)
			}
		}
	}

	fun indexOf(item: PaletteItem): Int {
		return items.indexOfFirst { it.id == item.id }
	}

	override fun getItemCount(): Int = items.size

	override fun getItemId(position: Int): Long {
		// TODO: use string to long id mapper (store and fetch via controller)
		return items[position].id.hashCode().toLong()
	}

	// TODO: rename to PaletteItemViewHolder
	class ColorViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		val outline: ImageView = itemView.findViewById(R.id.outline)
		val background: ImageView = itemView.findViewById(R.id.background)
	}
}